### PR TITLE
Mesh refactoring

### PR DESCRIPTION
**Pull request by biancagi**
Moved support for boundary calculation outside of class `Mesh`. 
Added mesh subcells filtering capability.
Improved interface of `SegementedVector`, which is now more consistent with the vector class from the standard library.

### CODE REVIEW

File: lib/mito/mesh/api.h [hunk 1]
```diff
@@ -13,6 +13,28 @@ namespace mito::mesh {
     // mesh factory
     template <class cellT>
     auto mesh() -> mesh_t<cellT>;
+
+
```
Comment (akoshakji) [inline lib/mito/mesh/api.h:17]: Remove extra new line.

File: lib/mito/utilities/SegmentedAllocator.h [hunk 2]
```diff
@@ -291,6 +342,8 @@ namespace mito::utilities {
         int _n_elements;
         // a queue with the available locations for writing
         std::queue<unqualified_pointer> _available_locations;
+        // a map with pointers to the start of each segment
+        std::map<int, unqualified_pointer> _start_of_segment;
```
Comment (akoshakji) [inline lib/mito/utilities/SegmentedAllocator.h:346]: Can be `std::unordered_map`?

File: lib/mito/utilities/SegmentedVector.h [hunk 3]
```diff
@@ -54,9 +54,23 @@ namespace mito::utilities {
             }
         }
 
+        // const components accessor (random access, may return an invalid resource)
+        inline auto operator[](int i) -> resource_type &
+        {
+            // delegate the answer to the allocator
+            return _resources[i];
```
Comment (akoshakji) [inline lib/mito/utilities/SegmentedVector.h:61]: This is risky. Might be good to do some boundary checks.

File: lib/mito/utilities/SegmentedVector.h [hunk 4]
```diff
@@ -54,9 +54,23 @@ namespace mito::utilities {
...
+        }
+
+        // components accessor (random access, may return an invalid resource)
+        inline auto operator[](int i) const -> const resource_type &
+        {
+            // delegate the answer to the allocator
+            return _resources[i];
```
Comment (akoshakji) [inline lib/mito/utilities/SegmentedVector.h:68]: Ditto.

File: lib/mito/utilities/SegmentedVector.h [hunk 5]
```diff
@@ -65,21 +79,25 @@ namespace mito::utilities {
             resource_type * resource = new (location) resource_type(std::forward<Args>(args)...);
 
             // add resource to the collection of resources
-            return iterator(_resources.insert(resource));
+            _resources.insert(resource);
```
Comment (akoshakji) [inline lib/mito/utilities/SegmentedVector.h:82]: Is this guaranteed to succeed? Or is it checked?

File: tests/mito.lib/utilities/segmented_vector_subscript.cc [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    // emplace resource in the container
+    collection.emplace(5);
+    EXPECT_TRUE(collection[5].foo() == 5);
+
+    // erase a resource
+    collection.erase(resource);
+
+    // check that the resource erased has been marked as invalid
+    EXPECT_TRUE(collection[2].is_valid() == false);
```
Comment (akoshakji) [inline tests/mito.lib/utilities/segmented_vector_subscript.cc:56]: `EXPECT_FALSE`?

File: tests/mito.lib/utilities/segmented_vector_subscript.cc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    collection.emplace(3);
+    EXPECT_TRUE(collection[3].foo() == 3);
+
+    // emplace resource in the container
+    collection.emplace(4);
+    EXPECT_TRUE(collection[4].foo() == 4);
+
+    // emplace resource in the container
+    collection.emplace(5);
+    EXPECT_TRUE(collection[5].foo() == 5);
```
Comment (akoshakji) [inline tests/mito.lib/utilities/segmented_vector_subscript.cc:50]: `EXPECT_EQ`?

**Review by akoshakji — COMMENTED** [review]
LGTM. Minor comments for discussion.

File: lib/mito/utilities/SegmentedVector.h (hunk 3 above)
Comment (biancagi) [reply lib/mito/utilities/SegmentedVector.h:61]: I would leave the `operator[]` as a true wrapper to `operator[]` and perhaps implement `at` with bounds checking, for consistency with the STL vector.

**Review by akoshakji — APPROVED** [review]

**Pull request merged**